### PR TITLE
Refactor loading weights

### PR DESCRIPTION
**Pull request by grimoire**
optimization tp model loading.

## requirement
- [x] #1520

### CODE REVIEW

File: lmdeploy/pytorch/weight_loader/model_weight_loader.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+
+def _get_weight_map(model_path: str, weight_type: str):
+    """get weight index."""
+    if weight_type == 'safetensors':
+        load_index = osp.join(model_path, SAFE_WEIGHTS_INDEX_NAME)
+    elif weight_type == 'pytorch':
+        load_index = osp.join(model_path, WEIGHTS_INDEX_NAME)
+    else:
+        raise RuntimeError('Unknown weight type.')
```
Comment (RunningLeon) [inline lmdeploy/pytorch/weight_loader/model_weight_loader.py:52]: may include `weight_type` in error message.

**Comment by RunningLeon** [comment]
@zhulinJulia24 hi, could you start a full-scope test of all pytorch engine models using `daily_test` CI? Thanks.

File: lmdeploy/pytorch/models/chatglm2.py [hunk 2]
```diff
@@ -91,52 +90,20 @@ def _distribute_qkv_linear(self, mod: nn.Module, device_mesh: DeviceMesh):
...
-            rowwise_parallelize_linear_fn(mod,
-                                          device_mesh=device_mesh,
-                                          to_local=True)
+        colwise_split_parallelize_linear(self.query_key_value,
+                                         sections,
+                                         loader,
+                                         rank=rank,
+                                         world_size=world_size,
+                                         prefix='query_key_value')
+        rowwise_parallelize_linear(self.dense,
```
Comment (lvhan028) [inline lmdeploy/pytorch/models/chatglm2.py:99]: much better than previous version

File: lmdeploy/pytorch/weight_loader/model_weight_loader.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+from .adapter_weight_loader import AdapterWeightLoader
+
+logger = get_logger('lmdeploy')
+
+
+def _get_weight_type(model_path: str, use_safetensors: bool = None):
+    """get weight type."""
+    weight_type = None
+    is_sharded = False
+    if use_safetensors is not False and osp.isfile(
```
Comment (lvhan028) [inline lmdeploy/pytorch/weight_loader/model_weight_loader.py:23]: `if user_safetensors is not False` -> `if user_safetensors`

File: lmdeploy/pytorch/weight_loader/model_weight_loader.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            osp.join(model_path, SAFE_WEIGHTS_NAME)):
+        # Load from a safetensors checkpoint
+        weight_type = 'safetensors'
+    elif use_safetensors is not False and osp.isfile(
```
Comment (lvhan028) [inline lmdeploy/pytorch/weight_loader/model_weight_loader.py:27]: `elif use_safetensors`

File: lmdeploy/pytorch/weight_loader/model_weight_loader.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+                                WEIGHTS_INDEX_NAME, WEIGHTS_NAME)
+
+from lmdeploy.utils import get_logger
+
+from .adapter_weight_loader import AdapterWeightLoader
+
+logger = get_logger('lmdeploy')
+
+
+def _get_weight_type(model_path: str, use_safetensors: bool = None):
```
Comment (lvhan028) [inline lmdeploy/pytorch/weight_loader/model_weight_loader.py:19]: `use_safetensors` can be {True, False, None}. Why not `True` or `False`?

Comment (grimoire) [reply lmdeploy/pytorch/weight_loader/model_weight_loader.py:19]: Align with transformers

https://github.com/huggingface/transformers/blob/dcdda5324bcc7a750b5e40e11dd795442204ff27/src/transformers/modeling_utils.py#L2813

File: lmdeploy/pytorch/weight_loader/model_weight_loader.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                             world_size: int = 1,
+                             device: torch.device = 'cpu'):
+    """load model weights implementation."""
+
+    def __load_no_recursive(mod: torch.nn.Module):
+        """load no recursive."""
+        for name, param in mod.named_parameters(recurse=False):
+            dtype = param.dtype
+            if not loader.has(name):
+                logger.debug(f'rank [{rank}]'
```
Comment (lvhan028) [inline lmdeploy/pytorch/weight_loader/model_weight_loader.py:170]: How to invoke this condition?

File: lmdeploy/pytorch/weight_loader/model_weight_loader.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+@torch.inference_mode()
+def load_model_weights(model: torch.nn.Module,
+                       checkpoint_path: str,
+                       adapters: Dict[str, str] = None,
+                       rank: int = 0,
+                       world_size: int = 1,
+                       device: torch.device = 'cpu'):
+    """Loading model weights.
+
+    Please waiting.
```
Comment (lvhan028) [inline lmdeploy/pytorch/weight_loader/model_weight_loader.py:215]: I didn't get "Please waiting"

File: lmdeploy/pytorch/models/functional.py [hunk 8]
```diff
@@ -160,204 +157,3 @@ def sync_qparam_to_context(context: Any, layer_id: str, qparams: dict):
         context.set_output(layer_id, last_qparam)
     else:
         context.set_output(layer_id, qparams)
-
-
-@torch.no_grad()
```
Comment (lvhan028) [inline lmdeploy/pytorch/models/functional.py:165]: Is it used before?

File: lmdeploy/pytorch/weight_loader/dist_utils.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+def colwise_split_parallelize_linear_naive(module: torch.nn.Module,
+                                           sections: List[int],
+                                           loader: ModelWeightLoader,
+                                           rank: int,
+                                           world_size: int,
+                                           prefix: str = ''):
+    """colwise split linear naive."""
+    for name, param in module.named_parameters():
+        dtype = param.dtype
```
Comment (lvhan028) [inline lmdeploy/pytorch/weight_loader/dist_utils.py:195]: It seems the for-named-parameters block is similar to the for-named-buffers block.
Can they be merged together?

File: lmdeploy/pytorch/weight_loader/dist_utils.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        module.register_buffer(name, param)
+
+
+def colwise_split_parallelize_loralinear(module: LoRALinear,
+                                         sections: List[int],
+                                         loader: ModelWeightLoader,
+                                         rank: int,
+                                         world_size: int,
+                                         prefix: str = ''):
+    """colwise split linear naive."""
```
Comment (lvhan028) [inline lmdeploy/pytorch/weight_loader/dist_utils.py:226]: docstring looks not right

File: lmdeploy/pytorch/weight_loader/model_weight_loader.py (hunk 6 above)
Comment (grimoire) [reply lmdeploy/pytorch/weight_loader/model_weight_loader.py:170]: Some model might shared weight of token embedding, they do not safe redundant weight in checkpoint.

File: lmdeploy/pytorch/models/functional.py (hunk 8 above)
Comment (grimoire) [reply lmdeploy/pytorch/models/functional.py:165]: Almost never.

**Review by lvhan028 — APPROVED** [review]

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**